### PR TITLE
KOGITO-2770: Changes on Setting Docks must activate isDirty status

### PR DESCRIPTION
**Pull request by yesamer**
**JIRA**: https://issues.redhat.com/browse/KOGITO-2770

**referenced Pull Requests**:
* https://github.com/kiegroup/drools/pull/3168

Screencast from 10-21-2020 02_42_46 PM

Aim of this PR is to handle any user change applied in Settings dock as a Undoable/Redoable action. This will enable the "isDirty" management in VScode. After any user change in SettingsDock, if user tries to close the editor, a confirmation popup will appear. This PR is valid for both Business Central and Kogito (VsCode)

Technical Description:
Before the change, Undoable commands belongs to Grid actions only. `AbstractScenarioSimulationGridCommand` managed this behavior. For the first time, this PR required to enable Undoable actions for a different widget than `GridPanel`. 
Therefore, I defined a new  `AbstractScenarioSimulationUndoableCommand`class in the commands hierarchy which manages the undoable / redoable behavior only. `AbstractScenarioSimulationGridCommand` now extends `AbstractScenarioSimulationUndoableCommand`, inheriting its behavior. New `UpdateSettingsDataCommands` extends `AbstractScenarioSimulationUndoableCommand` as well, managing the change on `Settings` only. The command contains a consumer (the change to apply to the model if valid) and the previous settings data status.

In addition: A lot of commands were set a CDI bean, even they are never used as Bean but only as Java objects (`new` is used)
I included minor fix for [DROOLS-5026](https://issues.redhat.com/browse/DROOLS-5026) as well

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>
 
* <b>a full downstream build</b> please add comment: <b>Jenkins run fdb</b>
  
* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by Rikkola** [comment]
@yesamer Was this issue just in kogito or also in business central?

**Comment by yesamer** [comment]
@Rikkola The changes will have impact on both BC and Kogito (VSCode). I'll provide a a full description later today.

**Review by Rikkola — APPROVED** [review]

**Comment by yesamer** [comment]
jenkins retest this please

**Review by jstastny-cz — CHANGES_REQUESTED** [review]

**Comment by jstastny-cz** [comment]
jenkins do fdb

**Comment by yesamer** [comment]
jenkins do fdb

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/ScenarioSimulationContext.java [hunk 1]
```diff
@@ -39,31 +39,23 @@
...
-import org.uberfire.mvp.PlaceRequest;
 
 /**
  * This class represent the <b>Context</b> inside which the commands will be executed
  */
 public class ScenarioSimulationContext {
 
-    private static final AtomicLong COUNTER_ID = new AtomicLong();
     private static final AtomicLong STATUS_COUNTER_ID = new AtomicLong();
+    private static final String ILLEGAL_WIDGET_MESSAGE = "Illegal GridWidget ";
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/ScenarioSimulationContext.java]: @jstastny-cz  Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/actualcommands/AbstractScenarioSimulationUndoableCommand.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        return commonUndoRedo(context);
+    }
+
+    public CommandResult<ScenarioSimulationViolation> redo(final ScenarioSimulationContext context) {
+        return commonUndoRedo(context);
+    }
+
+    private CommandResult<ScenarioSimulationViolation> commonUndoRedo(final ScenarioSimulationContext context) {
+        if (restorableStatus == null) {
+            String message = this.getClass().getSimpleName() + "restorableStatus status is null";
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/actualcommands/AbstractScenarioSimulationUndoableCommand.java]: @jstastny-cz  Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/actualcommands/UpdateSettingsDataCommand.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    @Override
+    public Optional<CommandResult<ScenarioSimulationViolation>> commonUndoRedoPreExecution(ScenarioSimulationContext context) {
+        context.getScenarioSimulationEditorPresenter().expandSettingsDock();
+        return Optional.of(CommandResultBuilder.SUCCESS);
+    }
+
+    protected CommandResult<ScenarioSimulationViolation> setCurrentContext(ScenarioSimulationContext context) {
+        try {
+            if (restorableStatus == null) {
+                throw new IllegalStateException("restorableSettings is null");
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/commands/actualcommands/UpdateSettingsDataCommand.java]: @jstastny-cz  Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/SettingsPresenter.java [hunk 4]
```diff
@@ -118,49 +123,78 @@ protected void setRuleSettings(Settings settings) {
...
     public void syncSkipFromBuild() {
-        settings.setSkipFromBuild(view.getSkipFromBuild().isChecked());
+        boolean isSkipFromBuild = view.getSkipFromBuild().isChecked();
+        eventBus.fireEvent(new UpdateSettingsDataEvent(settingsToUpdate -> settingsToUpdate.setSkipFromBuild(isSkipFromBuild)));
+    }
+
+    @Override
+    public void updateSettingsData(Settings settings) {
+        if (!isSettingTypeValid(settings.getType())) {
+            throw new IllegalStateException("Trying to update a " + settings.getType() + " data ");
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/SettingsPresenter.java]: @jstastny-cz  Done!

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/editor/ScenarioSimulationEditorPresenterTest.java [hunk 5]
```diff
@@ -645,7 +646,7 @@ public void getModelSuccessCallbackMethod() {
         presenterSpy.getModelSuccessCallbackMethod(dataManagementStrategyMock, modelLocal);
         verify(presenterSpy, times(1)).populateRightDocks(TestToolsPresenter.IDENTIFIER);
         verify(presenterSpy, times(1)).populateRightDocks(SettingsPresenter.IDENTIFIER);
-        verify(scenarioGridWidgetSpy, times(1)).setContent(eq(content.getModel().getSimulation()), eq(scenarioSimulationContextLocal.getSettings().getType()));
+        //verify(scenarioGridWidgetSpy, times(1)).setContent(eq(content.getModel().getSimulation()), eq(scenarioSimulationContextLocal.getStatus().getSettings().getType()));
```
[the remark this answers is not in the document]
Comment (yesamer) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/test/java/org/drools/workbench/screens/scenariosimulation/client/editor/ScenarioSimulationEditorPresenterTest.java]: @jstastny-cz  Done!

**Comment by jstastny-cz** [comment]
jenkins do fdb

**Comment by jstastny-cz** [comment]
Even though FDB failed the business central war was built successfully, continuing with testing.

**Comment by jstastny-cz** [comment]
BC - the editor is not locked when a change is done in settings panel. User does not have the feedback that the contents are eing edited by someone else until the other person saves the changes.

**Comment by yesamer** [comment]
@jstastny-cz Unfortunately the locking mechanism you mentioned works differently in BC. Considering this PR is mostly for KOGITO, I propose to manage the BC "isDirty" status in a different ticket. This because, if the PR works correctly in Kogito, I would prefer to don't block it.

**Comment by yesamer** [comment]
@jstastny-cz I tested the behavior in BC, and it seems correct to me. I opened a scesim file, I modified on parameter in Settings Dock and, if i try to leave without saving, a popup which warns the user of unsaved changes appears. If I undo the change and I close the editor, no message is shown as expected.

**Comment by jstastny-cz** [comment]
@yesamer yep that part works fine, just the locking does not work. I can state that as a non-goal for the feature and it would be ok to merge as is.

**Review by jstastny-cz — APPROVED** [review]
LGTM, we've come up to conclusion that the discussed behavior is in sync with what we have for other editors. Stating this explicitly as a non-goal for this feature, approving.

**Comment by yesamer** [comment]
@kiegroup/gatekeepers Can you please merge it?

**Pull request merged**